### ISSUE TITLE
Faircoin bug investigation results

### ISSUE THREAD

**Issue by bhaugen** [FreedomCoop/valuenetwork#197]
I'm posting these here now, but @santiky - who did the original work - may have more to add.

## Overview

This investigation was prompted by his finding of https://github.com/FreedomCoop/valuenetwork/issues/193 - where OCP was generating 10,000 unused faircoin accounts. 

After that, he ran into https://github.com/FreedomCoop/valuenetwork/issues/195 - where it is possible to get OCP to send a faircoin transaction without funds.

I raised another general issue about #193 - https://github.com/FreedomCoop/valuenetwork/issues/194 -  to limit the retries of all faircoin requests from OCP, so they do not keep trying forever after failing.

After #193 was bypassed by removing the offending unicode character, the faircoin address became the string "ERROR", which is also incorrect and will not please the OCP user who got that address. So I raised another issue, https://github.com/FreedomCoop/valuenetwork/issues/196 , about that problem.

## More about the thousands of faircoin addresses

Re #193 - this is the code that caused the problem. See the ```#**comments**``` on the right.

```
In daemon.py:
def new_fair_address(entity_id, entity = 'generic'):
    """ Return a new address labeled or False if there's no network connection. 
    The label is for debugging proposals. It's like 'entity: id'
    We can label like "user: 213" or "user: pachamama" or "order: 67".
    """
    while network.is_connected():
        new_address = wallet.create_new_address()      #**the address is created here**
        wallet.add_address(new_address)
        wallet.synchronizer.subscribe_to_addresses([new_address])
        wallet.up_to_date = False
        check_label = wallet.get_label(new_address)
        check_history = cmd_wallet.getaddresshistory(new_address)
        # It checks if address is labeled or has history yet, a gap limit protection. 
        # This can be removed when we have good control of gap limit.     
        if not check_label[0] and not check_history:
            wallet.set_label(new_address, entity + ': ' + str(entity_id))   #**error here**
            return new_address     #**so the new address was never sent back to OCP**
    return False
```

The error was caused by unicode characters in entity_id, which could not be handled by the str(entity_id) function. So I tried to fix that problem with [this commit](https://github.com/FreedomCoop/valuenetwork/commit/191bd17e8eb8140ee7c7bc03c66f2546e2ab3c0d), which changes the entity_id like this:
```
entity_id = agent.nick.encode('ascii','ignore'), 
```
That removes the unicode characters that cannot be handled by the str() function, which expects ascii.

However, other errors could also happen in the code after the address creation, which might also result in repeated bogus addresses that never make it back to OCP. So this new_fair_address function should be changed to either succeed completely and send a valid address back to OCP, or fail completely (that is, not leave a bogus new address in the wallet).

I added a note about that to the issue about [faircoin requests keep trying forever](https://github.com/FreedomCoop/valuenetwork/issues/194).

More testing will be required when the issues listed in this issue get fixed.

**Comment by bhaugen** [comment]
From @santiky in telegram: separate the address request-response from setting the label.

**Comment by bum2** [comment]
If @bhaugen can fix this issue will be wonderful... (tell me if you can't).
Another related thing pending to solve is: When we the devs copy the db in testocp to our local environments, and we also run our own electrum-fair local wallets, then the fairaddresses recorded in the db are not addresses of our local wallet (they're addresses of the testocp wallet)... I think is good at least to add a check when entering the account page to see if the recorded address pertains or not to the connected local wallet, and warn clearly about that. Perhaps just this now, and maybe a button to delete or recreate the fair address in a future... it only affects us the devs, but can lead to errors, isn't it?

**Comment by bhaugen** [comment]
> If @bhaugen can fix this issue will be wonderful... (tell me if you can't).

I can, maybe over the weekend. 

But will not deal with the dev environment wallet problem, which I do understand and experience. Maybe add another github issue for that one?
